paasta_tools/cleanup_kubernetes_cr.py

```diff
@@ -0,0 +1,142 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Usage: ./cleanup_kubernetes_cr.py [options]
+
+Command line options:
+
+- -d <SOA_DIR>, --soa-dir <SOA_DIR>: Specify a SOA config dir to read from
+- -v, --verbose: Verbose output
+"""
+import argparse
+import logging
+import sys
+from typing import Sequence
+
+from paasta_tools.kubernetes_tools import CustomResource
+from paasta_tools.kubernetes_tools import delete_custom_resource
+from paasta_tools.kubernetes_tools import KubeClient
+from paasta_tools.kubernetes_tools import list_custom_resources
+from paasta_tools.kubernetes_tools import load_custom_resources
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import load_all_configs
+from paasta_tools.utils import load_system_paasta_config
+
+log = logging.getLogger(__name__)
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description='Cleanup custom_resources.')
+    parser.add_argument(
+        '-d', '--soa-dir', dest="soa_dir", metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+    parser.add_argument(
+        '-v', '--verbose', action='store_true',
+        dest="verbose", default=False,
+    )
+    parser.add_argument(
+        '-c', '--cluster', default=None,
+        help="Cluster to cleanup CRs for",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def main() -> None:
+    args = parse_args()
+    soa_dir = args.soa_dir
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.INFO)
+
+    kube_client = KubeClient()
+
+    system_paasta_config = load_system_paasta_config()
+    cluster = args.cluster or system_paasta_config.get_cluster()
+    custom_resources = load_custom_resources(system_paasta_config)
+    cleanup_kube_succeeded = cleanup_all_custom_resources(
+        kube_client=kube_client,
+        soa_dir=soa_dir,
+        cluster=cluster,
+        custom_resources=custom_resources,
+    )
+    sys.exit(0 if cleanup_kube_succeeded else 1)
+
+
+def cleanup_all_custom_resources(
+    kube_client: KubeClient,
+    soa_dir: str,
+    cluster: str,
+    custom_resources: Sequence[CustomResource],
+) -> bool:
+    cluster_crds = {
+        crd.spec.names.kind
+        for crd in
+        kube_client.apiextensions.list_custom_resource_definition(
+            label_selector="yelp.com/paasta_service",
+        ).items
```

Comment:
doesn't this need to be `items()` ?

Reply:
it is `items` since we need that to work both with a list and with a mock:
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

kube_client = <Mock id='139930041401472'>, soa_dir = '/nail/soa', cluster = <Mock id='139930041401640'>
custom_resource_definitions = [<Mock id='139930035160736'>, <Mock id='139930035162304'>], service = None, instance = None

    def setup_all_custom_resources(
        kube_client: KubeClient,
        soa_dir: str,
        cluster: str,
        custom_resource_definitions: Sequence[CustomResourceDefinition],
        service: str = None,
        instance: str = None,
    ) -> bool:
        cluster_crds = {
            crd.spec.names.kind
            for crd in
            kube_client.apiextensions.list_custom_resource_definition(
>               label_selector="yelp.com/paasta_service",
            )
        }
E       TypeError: 'Mock' object is not iterable

paasta_tools/setup_kubernetes_cr.py:155: TypeError
```

Reply:
ohhh, sorry, `items` is actually a property on that thing, not just the regular `items()` func on a dict. ignore me.